raiden/transfer/channel.py

```diff
@@ -602,6 +612,26 @@ def get_known_unlocks(end_state: NettingChannelEndState) -> typing.List[UnlockPr
     ]
 
 
+def get_known_unlocks2(end_state: NettingChannelEndState) -> UnlockProofState2:
+    """ Unlock proof for an entire merkle tree of pending locks
+
+    The unlock proof contains all the merkle tree data, tightly packed, needed by the token
+    network contract to verify the secret expiry and calculate the token amounts to transfer.
+    """
+
+    all_locks = end_state.secrethashes_to_lockedlocks.copy()
+    all_locks.update({
+        lockhash: end_state.secrethashes_to_unlockedlocks[lockhash].lock
+        for lockhash in end_state.secrethashes_to_unlockedlocks.keys()
+    })
+
+    all_locks_packed = b''.join([
+        all_locks[lockhash].encoded for lockhash in end_state.merkletree.layers[LEAVES]
+    ])
+
+    return UnlockProofState2(all_locks_packed)
```

Comment:
The representation of the locks is really only important for the outer system, and not for the business logic. I think this is best done at the edge of the code base (i.e. the raiden event handler), instead of the core logic, since this is more of a detail on how the systems communicate than a business rule.